### PR TITLE
Issue #191 fixes both $..allTheThings.max() and $.max($..allTheThings)

### PR DESCRIPTION
**Pull request by mgreenwood1001**
Issue #191 is an unintentional side-effect of the collection processing for parameters passed to functions.  This change supports result set functions via patterns such as:

$.max($..allTheNumberThings)
    
$.max($.foo, $.bar) is already supported where foo and bar are numbers - consequently 

$.max($..allTheNumberThings) shall also work also 

$.avg($..allTheNumberThings, $.someCollection.max(), $..someOtherNumberThings)
will work too now with this change.

### CODE REVIEW

File: json-path/src/test/java/com/jayway/jsonpath/internal/function/Issue191.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * TDD for Issue 191
+ *
+ * Shows aggregation across fields rather than within a single entity.
+ *
+ */
+public class Issue191 extends com.jayway.jsonpath.internal.function.BaseFunctionTest {
```
Comment (mgreenwood1001) [inline json-path/src/test/java/com/jayway/jsonpath/internal/function/Issue191.java]: IntelliJ was having a bad day I guess, its in the same package, ugh - I'll fix this...

**Comment by mgreenwood1001** [comment]
the original question raised by #191 wanted to use the notation:

$..pathWithResultSetOutput.function()

so I've added support for that as well.

Normally, this won't work unless you make the graph stateful or introduce a pre-aggregation function (or emulate this in the evaluation steps) -- aggregate the results as a collection and passes it along to the function, i.e.  $..path.group().min() -- I'm not fond of this approach.

Instead, The CompilePath function now recognizes the condition where you're scanning the elements and introducing a collection as a the result set and attempting to pass that along as input into the function, since that won't work (functions are stateless and require parameters that can be lazy evaluated or passed) the CompilePath now inverts the graph making the function the root token and the parameter to the function the original path which makes the whole thing sane such that we can lazy evaluate the path get the result then evaluate the function.

So now:

$..path.sum() and

$.sum($..path) are supported, it won't matter which you specify as both result in the same graph internally.

Since this CompilePath trick is in the constructor, its recursive, so 

$.sum($..path.sum(), $..path.avg())

do what you'd hope it would do...

**Comment by rhavermans** [comment]
@kallestenflo Could you make a new release? We would like to have this functionality.

**Comment by soberich** [comment]
@mgreenwood1001 @kallestenflo  Hey, guys, I just stumbled upon fact that this is really weird.. Why is it many useful commits are on master but still was not released for more than a year? Anyone?

**Comment by mgreenwood1001** [comment]
@kallestenflo should you need anything from me for a release please let me know.

**Comment by soberich** [comment]
I've made pr with `div()` and `mult()` functions https://github.com/json-path/JsonPath/pull/511

**Comment by sarod** [comment]
@kallestenflo @mgreenwood1001 this feature would be really useful. It was merged in a year and a half ago. Is there a release planned soon?

**Comment by aabrams** [comment]
+1 Really useful to have this released, thanks!

**Comment by ericnesschrw** [comment]
This would be really useful. Any plans to release it?

**Pull request merged**